File: negotiation-cordapp/contracts-java/src/test/java/negotiation/contracts/ProposalContractTests.java
package negotiation.contracts;

import com.google.common.collect.ImmutableList;
import negotiation.states.ProposalState;
import net.corda.testing.contracts.DummyState;
import net.corda.testing.core.DummyCommandData;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;


import java.time.Instant;

import static net.corda.testing.node.NodeTestUtils.ledger;

public class ProposalContractTests {

    private MockServices ledgerServices = new MockServices(ImmutableList.of("negotiation.contracts"));
    private TestIdentity alice = new TestIdentity(new net.corda.core.identity.CordaX500Name("alice","New York", "US"));
    private TestIdentity bob = new TestIdentity(new net.corda.core.identity.CordaX500Name("bob","Tokyo", "JP"));
    private TestIdentity charlie = new TestIdentity(new net.corda.core.identity.CordaX500Name("charlie","London", "GB"));

    @Test
    public void proposalTransactionsHaveOnlyOneOutputOfTypeProposalState(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                tx.tweak(tdsl->{
                    tdsl.output(ProposalAndTradeContract.ID,new DummyState());
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(2,alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(2,alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.output(ProposalAndTradeContract.ID, new ProposalState(2,alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void proposalTransactionsHaveExactlyOneCommandOfTypePropose(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.tweak(tdsl -> {
                    tdsl.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()),  DummyCommandData.INSTANCE);
                    tdsl.fails();
                    return null;
                });
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                tx.verifies();
                return null;
            });
            return null;
        });
    }



    @Test
    public void proposalTransactionsHaveTwoRequiredSignersTheProposerAndProposee(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.tweak(tdsl ->{
                    tdsl.command(ImmutableList.of(alice.getPublicKey(), charlie.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.command(ImmutableList.of(charlie.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                    tdsl.fails();
                    return null;
                });
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                tx.verifies();
                return null;
            });
            return null;
        });
    }

    @Test
    public void inProposalTransactionsProposerAndProposeeAreBuyerAndSeller(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                tx.tweak(tdsl ->{
                    //order reversed - buyer = proposee, seller = proposer
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), bob.getParty(), alice.getParty()));
                    tdsl.verifies();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(2, charlie.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl -> {
                    tdsl.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), charlie.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.verifies();
                return null;
            });
            return null;
        });
    }



    @Test
    public void proposalAcceptanceTransactionsHaveNoInputsAndNoTimeStamp(){
        ledger(ledgerServices, ledger ->{
            ledger.transaction(tx ->{
                tx.output(ProposalAndTradeContract.ID, new ProposalState(2, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                tx.command(ImmutableList.of(alice.getPublicKey(), bob.getPublicKey()), new ProposalAndTradeContract.Commands.Propose());
                tx.tweak(tdsl ->{
                    tdsl.input(ProposalAndTradeContract.ID, new ProposalState(1, alice.getParty(), bob.getParty(), alice.getParty(), bob.getParty()));
                    tdsl.fails();
                    return null;
                });
                tx.tweak(tdsl ->{
                    tdsl.timeWindow(Instant.now());
                    tdsl.fails();
                    return null;
                });
                return null;
            });
            return null;
        });
    }

}

File: negotiation-cordapp/contracts-kotlin/src/test/kotlin/negotiation/contracts/ProposalContractTests.kt
package negotiation.flows.contracts

import negotiation.contracts.ProposalAndTradeContract
import negotiation.states.ProposalState
import net.corda.core.identity.CordaX500Name
import net.corda.testing.contracts.DummyContract
import net.corda.testing.contracts.DummyState
import net.corda.testing.core.DummyCommandData
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import org.junit.Test
import java.time.Instant

class ProposalContractTests {
    private val ledgerServices = MockServices(listOf("negotiation.contracts", "net.corda.testing.contracts"))
    private val alice = TestIdentity(CordaX500Name("alice", "New York", "US"))
    private val bob = TestIdentity(CordaX500Name("bob", "Tokyo", "JP"))
    private val charlie = TestIdentity(CordaX500Name("charlie", "London", "GB"))

    @Test
    fun `proposal transactions have exactly one output of type ProposalState`() {
        ledgerServices.ledger {
            transaction {
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                fails()
                tweak {
                    output(ProposalAndTradeContract.ID, DummyState())
                    fails()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                    fails()
                }
                output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                verifies()
            }
        }
    }

    @Test
    fun `proposal transactions have exactly one command of type Propose`() {
        ledgerServices.ledger {
            transaction {
                output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                tweak {
                    command(listOf(alice.publicKey, bob.publicKey), DummyCommandData)
                    fails()
                }
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                verifies()
            }
        }
    }

    @Test
    fun `proposal transactions have two required signers - the proposer and the proposee`() {
        ledgerServices.ledger {
            transaction {
                output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                tweak {
                    command(listOf(alice.publicKey, charlie.publicKey), ProposalAndTradeContract.Commands.Propose())
                    fails()
                }
                tweak {
                    command(listOf(charlie.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                    fails()
                }
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                verifies()
            }
        }
    }

    @Test
    fun `in proposal transactions, the proposer and proposee are the buyer and seller`() {
        ledgerServices.ledger {
            transaction {
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                tweak {
                    // Order reversed - buyer = proposee, seller = proposer
                    output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, bob.party, alice.party))
                    verifies()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, ProposalState(1, charlie.party, bob.party, alice.party, bob.party))
                    fails()
                }
                tweak {
                    output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, charlie.party, alice.party, bob.party))
                    fails()
                }
                output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                verifies()
            }
        }
    }

    @Test
    fun `proposal transactions have no inputs and no timestamp`() {
        ledgerServices.ledger {
            transaction {
                output(ProposalAndTradeContract.ID, ProposalState(1, alice.party, bob.party, alice.party, bob.party))
                command(listOf(alice.publicKey, bob.publicKey), ProposalAndTradeContract.Commands.Propose())
                tweak {
                    input(DummyContract.PROGRAM_ID, DummyState())
                    fails()
                }
                tweak {
                    timeWindow(Instant.now())
                    fails()
                }
            }
        }
    }
}